drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2024 ANITRA system s.r.o.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microcrystal_rv3028_retmem
+
+#include <zephyr/drivers/retained_mem.h>
+#include "rtc_rv3028.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(rv3028, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#define RV3028_USER_RAM_SIZE (2 * sizeof(uint8_t))
+
+struct rv3028_retained_mem_config {
+	const struct device *parent;
+};
+
+static ssize_t rv3028_retained_mem_size(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return RV3028_USER_RAM_SIZE;
+}
+
+static int rv3028_retained_mem_read(const struct device *dev, off_t offset, uint8_t *buffer,
+				    size_t size)
+{
+	const struct rv3028_retained_mem_config *config = dev->config;
+	const struct device *rtc_dev = config->parent;
+
+	return rv3028_read_regs(rtc_dev, (RV3028_REG_USER_RAM1 + offset), buffer, size);
+}
+
+static int rv3028_retained_mem_write(const struct device *dev, off_t offset, const uint8_t *buffer,
+				     size_t size)
+{
+	const struct rv3028_retained_mem_config *config = dev->config;
+	const struct device *rtc_dev = config->parent;
+
+	return rv3028_write_regs(rtc_dev, (RV3028_REG_USER_RAM1 + offset), buffer, size);
+}
+
+static int rv3028_retained_mem_clear(const struct device *dev)
+{
+	const struct rv3028_retained_mem_config *config = dev->config;
+	const struct device *rtc_dev = config->parent;
+	uint8_t block[RV3028_USER_RAM_SIZE] = {0};
+
+	return rv3028_write_regs(rtc_dev, RV3028_REG_USER_RAM1, block, RV3028_USER_RAM_SIZE);
+}
+
+static int rv3028_retmem_init(const struct device *dev)
+{
+	const struct rv3028_retained_mem_config *config = dev->config;
+
+	if (!device_is_ready(config->parent)) {
+		LOG_ERR("parent device %s is not ready", config->parent->name);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static const struct retained_mem_driver_api rv3028_retmem_api = {
+	.size = rv3028_retained_mem_size,
+	.read = rv3028_retained_mem_read,
+	.write = rv3028_retained_mem_write,
+	.clear = rv3028_retained_mem_clear,
+};
+
+#define RV3028_RETMEM_DEFINE(inst)                                                                 \
+	static const struct rv3028_retained_mem_config rv3028_retained_mem_config_##inst = {       \
+		.parent = DEVICE_DT_GET(DT_INST_BUS(inst)),                                        \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, rv3028_retmem_init, NULL, NULL,                                \
+			      &rv3028_retained_mem_config_##inst, POST_KERNEL,                     \
+			      CONFIG_RTC_INIT_PRIORITY, &rv3028_retmem_api);
```

Comment:
`CONFIG_RETAINED_MEM_INIT_PRIORITY`

Reply:
Using the parent (RTC) init priority was cautiously suggested by @fabiobaltieri, and I like this approach despite it potentially looking odd. The RV3028 is an I2C RTC, hence it is initialized relatively late by default. I don’t want to require users to push the retmem init priority past I2C and RTC, as there could be other retmem devices that do not depend on other components.

Reply:
this is a retained memory driver, it needs to use the retention init priority. Code (application and driver) should know when things init, it makes no sense and only provides mass confusion if someone wants to use this but needs to use it earlier if they reduce the retained memory init priority, use it, and just undefined operation and then have to spend a long time trying to work out what's going wrong when the issue is caused by a driver class using what is basically a completely random init priority that it should not be using. The nodes should be init'd in order, there is a python script which will check the priorities so if someone sets retained memory lower than the rtc the script should throw an error (can try that in a sample build and see)

Reply:
it's not a random init priority. it's an init priority of the actual device driver. this is not a standalone retmem device. this is an RTC implementing a retmem interface.

if this causes confusion as you suggest, I would revert to the previous solution - a custom init priority symbol (`CONFIG_RETAINED_MEM_RV3028_INIT_PRIORITY`), that would be visible from menuconfig

Reply:
Reverted to the custom init symbols